pkg/provision/storage/perWorkspaceStorage.go

```diff
@@ -170,10 +173,49 @@ func syncPerWorkspacePVC(workspace *common.DevWorkspaceWithConfig, clusterAPI sy
 		}
 	}
 
-	storageClass := workspace.Config.Workspace.StorageClassName
-	pvc, err := getPVCSpec(common.PerWorkspacePVCName(workspace.Status.DevWorkspaceId), workspace.Namespace, storageClass, pvcSize)
+	pvc := getPVCSpec(common.PerWorkspacePVCName(workspace.Status.DevWorkspaceId), workspace.Namespace, workspace.Config.Workspace.StorageClassName, pvcSize)
+
+	namespaceName := types.NamespacedName{
+		Name:      common.PerWorkspacePVCName(workspace.Status.DevWorkspaceId),
+		Namespace: workspace.Namespace,
+	}
+
+	existingPVC := &corev1.PersistentVolumeClaim{}
+	err = clusterAPI.Client.Get(clusterAPI.Ctx, namespaceName, existingPVC)
+
 	if err != nil {
-		return nil, err
+		if !k8sErrors.IsNotFound(err) {
+			return nil, err
+		}
+	} else if workspace.Config.Workspace.DefaultStorageSize.PerWorkspace != nil && PVCExpansionRequired(*existingPVC, workspace.Config.Workspace.DefaultStorageSize.PerWorkspace) && *workspace.Config.EnableExperimentalFeatures {
+		canExpand, err := StorageClassAllowsVolumeExpansion(workspace.Namespace, *existingPVC.Spec.StorageClassName, clusterAPI)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if canExpand {
+			// Resize the PVC
+			existingPVC.Spec.Resources.Requests[corev1.ResourceStorage] = *workspace.Config.Workspace.DefaultStorageSize.PerWorkspace
+			pvc = existingPVC
+
+			err = clusterAPI.Client.Update(clusterAPI.Ctx, pvc)
+			if err != nil {
+				return nil, err
+			}
+
+			msg, err := CheckPVCEvents(*pvc, clusterAPI)
+
+			if err != nil {
+				return nil, err
+			}
+
+			if msg != "" {
+				// TODO: Not sure if this is the best way to log in this case
+				// TODO: The error message could be better? Currently, it'll be something like: "PVC expansion error: Unable to expand PVC. More information: ..."
+				log.Log.Info(fmt.Sprintf("PVC expansion error: %s", msg))
```

Comment:
```suggestion
				clusterAPI.Logger.Info(fmt.Sprintf("PVC expansion error: %s", msg))
```
Though this may result in logging repeatedly (on every reconcile), which would fill up operator logs needlessly. Do we expect transient errors like this? Could this be better suited to a workspace warning condition?

Reply:
IIRC this repeated logging is part of the reason why I've been skeptical of this PR/feature (in its current state). I believe repeated logging here would happen in the following case:

> If a ResourceQuota or LimitRange prevents the PVC expansion request, then an error will be returned and continuously logged by DWO.

A workspace warning condition would be much better, though I have to figure out the "warning return flow" for this case.